### PR TITLE
Add CuBuffer and CuView

### PR DESCRIPTION
**Pull request by multitalentloes**
Add GPU types that will support more arbitrary buffers (vectors without the mathematical functionality of CuVector), and CuView, which stores pointers to memory typically stored on GPUs that will enable GPU instantiation of for instance parameter objects in the property evaluation later.

### CODE REVIEW

File: opm/simulators/linalg/cuistl/CuBuffer.hpp [hunk 1]
```diff
@@ -0,0 +1,248 @@
+/*
+  Copyright 2022-2023 SINTEF AS
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/CuBuffer.hpp:2]: A new file would presumably have copyright 2024?

File: opm/simulators/linalg/cuistl/CuBuffer.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+    : m_numberOfElements(detail::to_int(numberOfElements))
+{
+    OPM_CUDA_SAFE_CALL(cudaMalloc(&m_dataOnDevice, sizeof(T) * detail::to_size_t(m_numberOfElements)));
+}
+
+template <class T>
+CuBuffer<T>::CuBuffer()
+    : m_numberOfElements(0)
+{
+    OPM_CUDA_SAFE_CALL(cudaMalloc(&m_dataOnDevice, 0));
```
Comment (atgeirr) [inline opm/simulators/linalg/cuistl/CuBuffer.cpp:48]: This is weird... What happens when you allocate 0 bytes? Is it not better to just keep m_dataOnDevice a null pointer?

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 3]
```diff
@@ -0,0 +1,394 @@
...
+ *  inside a kernel to avoid expensive RDC (relocatable device code)
+ *
+ * The view will typically provide a view into a CuBuffer and be able to
+ * manipulate the data within it
+ *
+ * @param T Type of the data we store, typically int/float/double w/o const specifier
+ *
+ **/
+template <typename T>
+struct CuView
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:50]: Should be a `class` and should have its values (`m_dataPtr` etc etc) private.

File: opm/simulators/linalg/cuistl/CuBuffer.cpp (hunk 2 above)
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/CuBuffer.cpp:48]: Just scratch this base case constructor? Local tests indicate that this works, but I agree it looks dubious.

File: opm/simulators/linalg/cuistl/CuView.hpp (hunk 3 above)
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/CuView.hpp]: done

File: opm/simulators/linalg/cuistl/CuBuffer.hpp (hunk 1 above)
Comment (multitalentloes) [reply opm/simulators/linalg/cuistl/CuBuffer.hpp]: fixed

**Comment by multitalentloes** [comment]
Jenkins build this please

**Comment by multitalentloes** [comment]
This has been tested to work with PiecewiseLinearTwoPhaseMaterialParams, where one would now be able to instantiate these properties with CuBuffers and make a  view([#4081](https://github.com/OPM/opm-common/pull/4081)) that makes linear interpolation accessible on the GPU.

**Comment by multitalentloes** [comment]
Jenkins build this please

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+
+    //TODO: we probably dont need anything like this or is it useful to have views also be able to handle things on CPU?
+    /// @brief constructor based on std::vectors, this will make a view on the CPU
+    /// @param data std vector to pr
+    CuView(std::vector<T>& data);
+
+    /**
+     * @brief operator[] to retrieve a reference to an item in the buffer
+     *
+     * @note This does asynchronous operations
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:66]: What asynchronous operation?

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+#ifndef NDEBUG
+    assertInRange(idx);
+#endif
+    return m_dataPtr[idx];
+    }
+
+    /**
+     * @brief operator[] to retrieve a copy of an item in the buffer
+     *
+     * @note This does asynchronous operations
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:80]: What asynchronous operation?

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 6]
```diff
@@ -0,0 +1,395 @@
...
+#endif
+    return m_dataPtr[m_numberOfElements-1];
+    }
+
+    /**
+     * @return fetch the first element in a CuView
+     */
+    __host__ __device__ T front() const
+    {
+#ifndef NDEBUG
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:152]: rest of function: Indentation

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+#endif
+    return m_dataPtr[0];
+    }
+
+    /**
+     * @return fetch the last element in a CuView
+     */
+    __host__ __device__ T& back()
+    {
+#ifndef NDEBUG
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:141]: rest of function: Indentation

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+    assertHasElements();
+#endif
+    return m_dataPtr[0];
+    }
+
+    /**
+     * @return fetch the last element in a CuView
+     */
+    __host__ __device__ T back() const
+    {
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:162]: rest of function: Indentation

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 9]
```diff
@@ -0,0 +1,395 @@
...
+     * @note This assumes that the size of this vector is equal to the size of the input vector.
+     */
+    void copyToHost(std::vector<T>& data) const;
+
+    /**
+     * @brief size returns the size (number of T elements) in the vector
+     * @return number of elements
+     */
+    __host__ __device__ size_t size() const{
+        // return std::size_t(m_numberOfElements);
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:210]: remove commented out line

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 10]
```diff
@@ -0,0 +1,395 @@
...
+        __host__ __device__ iterator operator--(int) {
+            iterator tmp = *this;
+            --m_ptr;
+            return tmp;
+        }
+
+        /// @brief Inequality comparison operator
+        /// @return boolean value that is true if the pointers contains different addresses
+        __host__ __device__ bool operator!=(const iterator& other) const {
+            return !(*m_ptr == *other.m_ptr);
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:275]: Why are you dereferencing here? The pointers can be unequal even though the value they point to is the same.

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 11]
```diff
@@ -0,0 +1,395 @@
...
+        }
+
+        /// @brief Inequality comparison operator
+        /// @return boolean value that is true if the pointers contains the same address
+        __host__ __device__ bool operator==(const iterator& other) const {
+            return *m_ptr == *other.m_ptr;
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:281]: again, why dereferencing?

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 12]
```diff
@@ -0,0 +1,395 @@
...
+        /// @param other iterator to subtract
+        /// @return diffptr that represents difference between the iterators
+        __host__ __device__ difference_type operator-(const iterator& other) const {
+            return std::distance(other.m_ptr, m_ptr);
+        }
+
+        /// @brief Subtraction of given number of elements from iterator
+        /// @param n the number of elements to step backwards
+        /// @return An iterator pointing to a location n steps behind
+        __host__ __device__ iterator operator-(int n) const {
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:294]: `difference_type`

**Review by kjetilly — CHANGES_REQUESTED** [review]
Overall good, some minor comments on the iterator functions, indentation and some documentation.

**Comment by multitalentloes** [comment]
Jenkins build this please

File: opm/simulators/linalg/cuistl/CuBuffer.cpp [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+
+template <class T>
+CuBuffer<T>::CuBuffer(const CuBuffer<T>& other)
+    : CuBuffer(other.m_numberOfElements)
+{
+    assertHasElements();
+    assertSameSize(other);
+    OPM_CUDA_SAFE_CALL(cudaMemcpy(m_dataOnDevice,
+                                  other.m_dataOnDevice,
+                                  detail::to_size_t(m_numberOfElements) * sizeof(T),
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.cpp:63]: `m_numberOfElements` is a `size_t` now, so `to_size_t` is not needed.

File: opm/simulators/linalg/cuistl/CuBuffer.cpp [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+CuBuffer<T>::~CuBuffer()
+{
+    OPM_CUDA_WARN_IF_ERROR(cudaFree(m_dataOnDevice));
+}
+
+template <typename T>
+typename CuBuffer<T>::size_type
+CuBuffer<T>::size() const
+{
+    return detail::to_size_t(m_numberOfElements);
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.cpp:77]: `m_numberOfElements` is a `size_t` now, so `to_size_t` is not needed.

File: opm/simulators/linalg/cuistl/CuBuffer.cpp [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+template <typename T>
+void
+CuBuffer<T>::resize(size_t newSize)
+{
+    if (newSize < 1) {
+        OPM_THROW(std::invalid_argument, "Setting a CuBuffer size to a non-positive number is not allowed");
+    }
+    // Allocate memory for the new buffer
+    T* tmpBuffer = nullptr;
+    OPM_CUDA_SAFE_CALL(cudaMalloc(&tmpBuffer, sizeof(T) * detail::to_size_t(newSize)));
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.cpp:89]: `newSize` is a `size_t` now, so `to_size_t` is not needed.

File: opm/simulators/linalg/cuistl/CuBuffer.cpp [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+    // Move the data from the old to the new buffer with truncation
+    size_t sizeOfMove = std::min({m_numberOfElements, newSize});
+    OPM_CUDA_SAFE_CALL(cudaMemcpy(tmpBuffer,
+                                  m_dataOnDevice,
+                                  detail::to_size_t(sizeOfMove) * sizeof(T),
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.cpp:95]: `sizeMove` is a `size_t` now, so `to_size_t` is not needed.

File: opm/simulators/linalg/cuistl/CuBuffer.cpp [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+
+    // update size
+    m_numberOfElements = newSize;
+}
+
+template <typename T>
+std::vector<T>
+CuBuffer<T>::asStdVector() const
+{
+    std::vector<T> temporary(detail::to_size_t(m_numberOfElements));
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.cpp:112]: `m_numberOfElements` is a `size_t` now, so `to_size_t` is not needed.

File: opm/simulators/linalg/cuistl/CuBuffer.hpp [hunk 18]
```diff
@@ -0,0 +1,280 @@
...
+    /**
+     * @return the raw pointer to the GPU data
+     */
+    const T* data() const;
+
+    /**
+     * @return fetch the first element in a CuBuffer
+     */
+    __host__ __device__ T& front()
+    {
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.hpp:126]: Indentation

File: opm/simulators/linalg/cuistl/CuBuffer.hpp [hunk 19]
```diff
@@ -0,0 +1,280 @@
...
+    assertHasElements();
+#endif
+    return m_dataOnDevice[0];
+    }
+
+    /**
+     * @return fetch the last element in a CuBuffer
+     */
+    __host__ __device__ T& back()
+    {
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.hpp:137]: Indentation

File: opm/simulators/linalg/cuistl/CuBuffer.hpp [hunk 20]
```diff
@@ -0,0 +1,280 @@
...
+    return m_dataOnDevice[m_numberOfElements-1];
+    }
+
+    /**
+     * @return fetch the first element in a CuBuffer
+     */
+    __host__ __device__ T front() const
+    {
+#ifndef NDEBUG
+    assertHasElements();
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.hpp:150]: Indentation

File: opm/simulators/linalg/cuistl/CuBuffer.hpp [hunk 21]
```diff
@@ -0,0 +1,280 @@
...
+    return m_dataOnDevice[0];
+    }
+
+    /**
+     * @return fetch the last element in a CuBuffer
+     */
+    __host__ __device__ T back() const
+    {
+#ifndef NDEBUG
+    assertHasElements();
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuBuffer.hpp:161]: Indentation

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 22]
```diff
@@ -0,0 +1,390 @@
...
+    CuView(std::vector<T>& data);
+
+    /**
+     * @brief operator[] to retrieve a reference to an item in the buffer
+     *
+     * @param idx The index of the element
+     */
+    __host__ __device__ T& operator[](size_t idx){
+#ifndef NDEBUG
+    assertInRange(idx);
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:70]: Indentation

File: opm/simulators/linalg/cuistl/CuView.cpp [hunk 23]
```diff
@@ -0,0 +1,82 @@
...
+#include <fmt/core.h>
+#include <opm/simulators/linalg/cuistl/CuView.hpp>
+#include <opm/simulators/linalg/cuistl/detail/cuda_safe_call.hpp>
+
+namespace Opm::cuistl
+{
+
+template <class T>
+CuView<T>::CuView(std::vector<T>& data)
+    : CuView(data.data(), detail::to_int(data.size()))
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.cpp:31]: `to_int` not needed anymore

File: opm/simulators/linalg/cuistl/CuView.cpp [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+                              numberOfElements));
+    }
+    OPM_CUDA_SAFE_CALL(cudaMemcpy(data(), dataPointer, numberOfElements * sizeof(T), cudaMemcpyHostToDevice));
+}
+
+template <class T>
+void
+CuView<T>::copyToHost(T* dataPointer, size_t numberOfElements) const
+{
+    assertSameSize(detail::to_int(numberOfElements));
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.cpp:61]: `to_int` not needed anymore.

File: opm/simulators/linalg/cuistl/CuView.hpp [hunk 25]
```diff
@@ -0,0 +1,390 @@
...
+    }
+
+    /**
+     * @brief operator[] to retrieve a copy of an item in the buffer
+     *
+     * @param idx The index of the element
+     */
+    __host__ __device__ T operator[](size_t idx) const {
+#ifndef NDEBUG
+    assertInRange(idx);
```
Comment (kjetilly) [inline opm/simulators/linalg/cuistl/CuView.hpp:82]: Indentation.

**Review by kjetilly — CHANGES_REQUESTED** [review]
Some dangling `size_t` / `int` conversions that are no longer needed.

**Comment by multitalentloes** [comment]
Jenkins build this please

**Comment by multitalentloes** [comment]
Jenkins build this please

**Review by kjetilly — APPROVED** [review]

**Pull request merged**